<a href="https://colab.research.google.com/github/keunhee-gom/DAT_Mainquest/blob/main/Feature_Engineering_project(1)_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 📋 프로젝트 개요

**목표**: 신용카드 사기거래를 탐지하기 위한 머신러닝 모델 데이터 전처리

**데이터**: fraud.csv (타겟 변수: `is_fraud`)

**핵심**: Feature Engineering을 통한 예측 성능 향상

In [250]:
import pandas as pd
import numpy as np
import seaborn as sns

In [251]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [252]:
cc_df= pd.read_csv('/content/drive/MyDrive/Colab Notebooks/pandas data/fraud.csv')

# **1. 데이터 정보 확인 및 전처리**

In [253]:
print(cc_df.shape)

(491134, 22)


In [254]:
pd.set_option('display.max_columns', 50)

In [255]:
cc_df.head()

trans_date_trans_time               cc_num  \
0   2019-01-01 00:00:44         630423337322   
1   2019-01-01 00:12:34  4956828990005111019   
2   2019-01-01 00:17:16      180048185037117   
3   2019-01-01 00:20:15      374930071163758   
4   2019-01-01 00:23:41     2712209726293386   

                            merchant     category     amt      first  \
0    fraud_Heller, Gutmann and Zieme  grocery_pos  107.23  Stephanie   
1  fraud_Schultz, Simonis and Little  grocery_pos   44.71    Kenneth   
2                  fraud_Kling-Grant  grocery_net   46.28       Mary   
3              fraud_Deckow-O'Conner  grocery_pos   64.09     Daniel   
4             fraud_Balistreri-Nader     misc_pos   25.58      Jenna   

       last gender                        street         city state    zip  \
0      Gill      F  43039 Riley Greens Suite 393       Orient    WA  99160   
1  Robinson      M            269 Sanchez Rapids    Elizabeth    NJ   7208   
2      Wall      F               2481 Mills Lock   Plainfield    NJ   7060   
3   Escobar      M              61390 Hayes Port      Romulus    MI  48174   
4    Brooks      F    50872 Alex Plain Suite 088  Baton Rouge    LA  70808   

       lat      long  city_pop                                job         dob  \
0  48.8878 -118.2105       149  Special educational needs teacher  1978-06-21   
1  40.6747  -74.2239    124967             Operational researcher  1980-12-21   
2  40.6152  -74.4150     71485             Leisure centre manager  1974-07-19   
3  42.2203  -83.3583     31515                     Police officer  1971-11-05   
4  30.4066  -91.1468    378909                Designer, furniture  1977-02-22   

                          trans_num   unix_time  merch_lat  merch_long  \
0  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
1  09eff9c806365e2a6be12c1bbab3d70e  1325376754  40.079588  -74.848087   
2  19e23c6a300c774354417befe4f31f8c  1325377036  40.021888  -74.228188   
3  6f363661ba6b55889e488dd178f2a0af  1325377215  42.360426  -83.552316   
4  1654da2abfb9e79a5f99167fc9779558  1325377421  29.737426  -90.853194   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0

In [256]:
cc_df['merchant'].nunique()

693

In [257]:
cc_df['category'].nunique()

14

In [258]:
cc_df['job'].nunique()

110

In [259]:
cc_df['cc_num'].nunique()

124

In [260]:
cc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 491134 entries, 0 to 491133
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   trans_date_trans_time  491134 non-null  object 
 1   cc_num                 491134 non-null  int64  
 2   merchant               491134 non-null  object 
 3   category               491134 non-null  object 
 4   amt                    491134 non-null  float64
 5   first                  491134 non-null  object 
 6   last                   491134 non-null  object 
 7   gender                 491134 non-null  object 
 8   street                 491134 non-null  object 
 9   city                   491134 non-null  object 
 10  state                  491134 non-null  object 
 11  zip                    491134 non-null  int64  
 12  lat                    491134 non-null  float64
 13  long                   491134 non-null  float64
 14  city_pop               491134 non-nu

### 📊 제거 이유

- **merchant**: 693개의 서로 다른 상점명 (카테고리로 충분)
- **개인정보**: first, last, street, city, state, zip (개인정보 보호)
- **중복정보**: trans_num, unix_time (trans_date_trans_time과 중복)

In [261]:
drop_columns = ['merchant', 'first', 'last', 'street', 'city', 'state',
                'zip', 'job', 'trans_num', 'unix_time']
cc_df.drop(drop_columns, axis=1, inplace=True)

In [262]:
cc_df.head()

trans_date_trans_time               cc_num     category     amt gender  \
0   2019-01-01 00:00:44         630423337322  grocery_pos  107.23      F   
1   2019-01-01 00:12:34  4956828990005111019  grocery_pos   44.71      M   
2   2019-01-01 00:17:16      180048185037117  grocery_net   46.28      F   
3   2019-01-01 00:20:15      374930071163758  grocery_pos   64.09      M   
4   2019-01-01 00:23:41     2712209726293386     misc_pos   25.58      F   

       lat      long  city_pop         dob  merch_lat  merch_long  is_fraud  
0  48.8878 -118.2105       149  1978-06-21  49.159047 -118.186462         0  
1  40.6747  -74.2239    124967  1980-12-21  40.079588  -74.848087         0  
2  40.6152  -74.4150     71485  1974-07-19  40.021888  -74.228188         0  
3  42.2203  -83.3583     31515  1971-11-05  42.360426  -83.552316         0  
4  30.4066  -91.1468    378909  1977-02-22  29.737426  -90.853194         0

In [263]:
cc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 491134 entries, 0 to 491133
Data columns (total 12 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   trans_date_trans_time  491134 non-null  object 
 1   cc_num                 491134 non-null  int64  
 2   category               491134 non-null  object 
 3   amt                    491134 non-null  float64
 4   gender                 491134 non-null  object 
 5   lat                    491134 non-null  float64
 6   long                   491134 non-null  float64
 7   city_pop               491134 non-null  int64  
 8   dob                    491134 non-null  object 
 9   merch_lat              491134 non-null  float64
 10  merch_long             491134 non-null  float64
 11  is_fraud               491134 non-null  int64  
dtypes: float64(5), int64(3), object(4)
memory usage: 45.0+ MB


# **2. Z-Score 계산**
## 1) 구매금액
각 고객의 평소 소비 패턴을 기준으로 이상 거래를 탐지. Z-Score가 높을수록 평소와 다른 거래패턴을 의미

In [264]:
# 고객별 구매금액 통계 계산
# 신용카드 번호(cc_num)별로 묶고, 거래 금액(amt) 컬럼 선택해서, 평균과 표준편차 계산
amt_info = cc_df.groupby('cc_num')['amt'].agg(['mean', 'std'])

In [265]:
amt_info.info()

<class 'pandas.core.frame.DataFrame'>
Index: 124 entries, 503874407318 to 4956828990005111019
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   mean    124 non-null    float64
 1   std     124 non-null    float64
dtypes: float64(2)
memory usage: 2.9 KB


In [266]:
amt_info.reset_index()

cc_num       mean         std
0           503874407318  60.253406  127.265783
1           567868110212  83.442558  117.303828
2           571365235126  59.392974  134.289959
3           581686439828  58.578675  149.804992
4           630423337322  56.078113  159.201852
..                   ...        ...         ...
119  4792627764422477317  84.135134  107.316736
120  4797297220948468262  56.313583  247.931817
121  4861310130652566408  85.805306  130.998089
122  4906628655840914250  54.243453  154.767184
123  4956828990005111019  59.858059  132.138802

[124 rows x 3 columns]

In [267]:
# amt_info 데이터프레임을 피클(pickle) 파일로 저장
# 저장된 파일 이름은 amt_info.pkl이고, 현재 작업 디렉토리에(./) 저장
amt_info.to_pickle('./amt_info.pkl')

In [268]:
# 메인 데이터와 통계 정보 병합
cc_df = cc_df.merge(amt_info,on='cc_num',how='left')

In [269]:
cc_df.head()

trans_date_trans_time               cc_num     category     amt gender  \
0   2019-01-01 00:00:44         630423337322  grocery_pos  107.23      F   
1   2019-01-01 00:12:34  4956828990005111019  grocery_pos   44.71      M   
2   2019-01-01 00:17:16      180048185037117  grocery_net   46.28      F   
3   2019-01-01 00:20:15      374930071163758  grocery_pos   64.09      M   
4   2019-01-01 00:23:41     2712209726293386     misc_pos   25.58      F   

       lat      long  city_pop         dob  merch_lat  merch_long  is_fraud  \
0  48.8878 -118.2105       149  1978-06-21  49.159047 -118.186462         0   
1  40.6747  -74.2239    124967  1980-12-21  40.079588  -74.848087         0   
2  40.6152  -74.4150     71485  1974-07-19  40.021888  -74.228188         0   
3  42.2203  -83.3583     31515  1971-11-05  42.360426  -83.552316         0   
4  30.4066  -91.1468    378909  1977-02-22  29.737426  -90.853194         0   

        mean         std  
0  56.078113  159.201852  
1  59.858059  132.138802  
2  87.328067  113.454416  
3  64.317839  174.739042  
4  90.747123  165.470881

In [270]:
# Z-Score 계산
# Z-Score = (개별 거래금액 - 고객 평균 거래금액) / 고객 거래금액 표준편차
# 평균에서 얼마나 떨어져 있는지를 표준편차 단위로 나타낸 값

cc_df['amt_z'] = (cc_df['amt']-cc_df['mean'])/cc_df['std']

In [271]:
cc_df['amt_z'].describe()

count    4.911340e+05
mean     1.793956e-18
std      9.998748e-01
min     -9.227851e-01
25%     -3.609422e-01
50%     -1.774582e-01
75%      9.572879e-02
max      6.342123e+01
Name: amt_z, dtype: float64

In [272]:
cc_df[cc_df['is_fraud']==1]

trans_date_trans_time            cc_num       category      amt gender  \
4794     2019-01-12 00:59:01      581686439828  gas_transport    11.73      M   
4816     2019-01-12 03:48:07      581686439828    grocery_pos   328.68      M   
4979     2019-01-12 15:46:10      581686439828    food_dining   120.58      M   
5073     2019-01-12 19:53:59      581686439828   shopping_net  1081.35      M   
5124     2019-01-12 22:44:05      581686439828   shopping_net   776.70      M   
...                      ...               ...            ...      ...    ...   
477832   2020-12-21 02:21:41  4716561796955522    grocery_pos   358.24      F   
477847   2020-12-21 02:36:03  4716561796955522   shopping_net   859.12      F   
479296   2020-12-21 22:38:38  4716561796955522           home   209.84      F   
479305   2020-12-21 22:42:11  4716561796955522    food_dining   123.58      F   
479341   2020-12-21 22:59:22  4716561796955522           home   219.11      F   

            lat      long  city_pop         dob  merch_lat  merch_long  \
4794    41.5205  -80.0573      5507  1973-07-28  41.947427  -79.796264   
4816    41.5205  -80.0573      5507  1973-07-28  42.148618  -79.398595   
4979    41.5205  -80.0573      5507  1973-07-28  42.470024  -80.126576   
5073    41.5205  -80.0573      5507  1973-07-28  42.455406  -79.521640   
5124    41.5205  -80.0573      5507  1973-07-28  40.680209  -79.099101   
...         ...       ...       ...         ...        ...         ...   
477832  48.2777 -112.8456       743  1972-05-04  47.526202 -113.643313   
477847  48.2777 -112.8456       743  1972-05-04  48.272348 -112.328075   
479296  48.2777 -112.8456       743  1972-05-04  49.173669 -112.698767   
479305  48.2777 -112.8456       743  1972-05-04  48.913048 -113.214921   
479341  48.2777 -112.8456       743  1972-05-04  47.911219 -113.691603   

        is_fraud       mean         std     amt_z  
4794           1  58.578675  149.804992 -0.312731  
4816           1  58.578675  149.804992  1.803020  
4979           1  58.578675  149.804992  0.413880  
5073           1  58.578675  149.804992  6.827351  
5124           1  58.578675  149.804992  4.793708  
...          ...        ...         ...       ...  
477832         1  52.537867  106.113023  2.880911  
477847         1  52.537867  106.113023  7.601161  
479296         1  52.537867  106.113023  1.482402  
479305         1  52.537867  106.113023  0.669495  
479341         1  52.537867  106.113023  1.569761  

[1244 rows x 15 columns]

In [273]:
print(cc_df.columns)

Index(['trans_date_trans_time', 'cc_num', 'category', 'amt', 'gender', 'lat',
       'long', 'city_pop', 'dob', 'merch_lat', 'merch_long', 'is_fraud',
       'mean', 'std', 'amt_z'],
      dtype='object')


In [274]:
# z-score 계산이 완료되었으니 mean, std 컬럼을 제거
cc_df.drop(['mean','std'], axis = 1, inplace = True)

In [275]:
print(cc_df.columns)

Index(['trans_date_trans_time', 'cc_num', 'category', 'amt', 'gender', 'lat',
       'long', 'city_pop', 'dob', 'merch_lat', 'merch_long', 'is_fraud',
       'amt_z'],
      dtype='object')


## 2) 카테고리별 구매 패턴
고객이 특정 카테고리에서 보이는 소비 패턴을 분석.
예를 들어, 평소 식료품만 구매하던 고객이 갑자기 고가의 전자제품을 구매하는 경우를 탐지

In [276]:
# 신용카드, 카테고리 별 구매 금액의 평균과 표준편차
cat_info = cc_df.groupby(['cc_num','category'])['amt'].agg(['mean','std']).reset_index()

In [277]:
cat_info

cc_num       category       mean         std
0            503874407318  entertainment  73.282418  103.050402
1            503874407318    food_dining  38.712305   46.548436
2            503874407318  gas_transport  68.457820   14.730440
3            503874407318    grocery_net  48.931302   18.736252
4            503874407318    grocery_pos  61.987806   23.449569
...                   ...            ...        ...         ...
1731  4956828990005111019       misc_pos  74.177012  168.341518
1732  4956828990005111019  personal_care  35.379382   44.082579
1733  4956828990005111019   shopping_net  70.019115  239.350164
1734  4956828990005111019   shopping_pos  45.988976  174.986921
1735  4956828990005111019         travel  72.701961  456.554619

[1736 rows x 4 columns]

In [278]:
cat_info.to_pickle('./cat_info.pkl')

In [279]:
cc_df = cc_df.merge(cat_info, on=['cc_num','category'], how='left')

In [280]:
cc_df.head()

trans_date_trans_time               cc_num     category     amt gender  \
0   2019-01-01 00:00:44         630423337322  grocery_pos  107.23      F   
1   2019-01-01 00:12:34  4956828990005111019  grocery_pos   44.71      M   
2   2019-01-01 00:17:16      180048185037117  grocery_net   46.28      F   
3   2019-01-01 00:20:15      374930071163758  grocery_pos   64.09      M   
4   2019-01-01 00:23:41     2712209726293386     misc_pos   25.58      F   

       lat      long  city_pop         dob  merch_lat  merch_long  is_fraud  \
0  48.8878 -118.2105       149  1978-06-21  49.159047 -118.186462         0   
1  40.6747  -74.2239    124967  1980-12-21  40.079588  -74.848087         0   
2  40.6152  -74.4150     71485  1974-07-19  40.021888  -74.228188         0   
3  42.2203  -83.3583     31515  1971-11-05  42.360426  -83.552316         0   
4  30.4066  -91.1468    378909  1977-02-22  29.737426  -90.853194         0   

      amt_z       mean         std  
0  0.321302  99.637224   23.904424  
1 -0.114637  60.694144   24.513316  
2 -0.361802  60.427269   19.558574  
3 -0.001304  59.145831   23.345746  
4 -0.393828  86.794025  106.330185

In [281]:
cc_df['cat_amt_z'] = (cc_df['amt'] -cc_df['mean'])/cc_df['std']

In [282]:
cc_df.head(10)

trans_date_trans_time               cc_num       category     amt gender  \
0   2019-01-01 00:00:44         630423337322    grocery_pos  107.23      F   
1   2019-01-01 00:12:34  4956828990005111019    grocery_pos   44.71      M   
2   2019-01-01 00:17:16      180048185037117    grocery_net   46.28      F   
3   2019-01-01 00:20:15      374930071163758    grocery_pos   64.09      M   
4   2019-01-01 00:23:41     2712209726293386       misc_pos   25.58      F   
5   2019-01-01 00:26:22     3598215285024754    grocery_pos  207.36      F   
6   2019-01-01 00:27:45      213141712584544       misc_pos  181.35      F   
7   2019-01-01 00:30:49         676372984911  gas_transport   53.60      F   
8   2019-01-01 00:31:51     4642255475285942  gas_transport  102.13      F   
9   2019-01-01 00:34:10     3553629419254918    grocery_pos   83.07      F   

       lat      long  city_pop         dob  merch_lat  merch_long  is_fraud  \
0  48.8878 -118.2105       149  1978-06-21  49.159047 -118.186462         0   
1  40.6747  -74.2239    124967  1980-12-21  40.079588  -74.848087         0   
2  40.6152  -74.4150     71485  1974-07-19  40.021888  -74.228188         0   
3  42.2203  -83.3583     31515  1971-11-05  42.360426  -83.552316         0   
4  30.4066  -91.1468    378909  1977-02-22  29.737426  -90.853194         0   
5  40.6729  -73.5365     34496  1970-10-21  40.202538  -73.859212         0   
6  32.1530  -90.1217     19685  1984-12-24  31.939812  -90.726209         0   
7  39.7562  -98.4462       242  1987-07-18  40.332519  -99.062962         0   
8  38.9999 -109.6150        46  1987-04-23  39.807313 -109.348294         0   
9  48.3400 -122.3456        85  1984-09-01  48.682111 -122.719904         0   

      amt_z        mean         std  cat_amt_z  
0  0.321302   99.637224   23.904424   0.317631  
1 -0.114637   60.694144   24.513316  -0.652060  
2 -0.361802   60.427269   19.558574  -0.723328  
3 -0.001304   59.145831   23.345746   0.211780  
4 -0.393828   86.794025  106.330185  -0.575698  
5  1.012426  200.841585   20.598234   0.316455  
6  0.766191   72.187689   91.470643   1.193414  
7 -0.035593   59.624852   15.552938  -0.387377  
8  0.296075   60.295358   14.952816   2.797777  
9  0.293265  101.476771   23.005191  -0.800114

In [283]:
cc_df.drop(['mean','std'], axis =1 , inplace = True)

In [284]:
cc_df.groupby('is_fraud')[['cat_amt_z','amt_z']].describe()
# 사기거래 Z-Score 평균이 카테고리별 금액의 경우 3,13, 고객별 금액의 경우 2.93로 정상거래보다 훨씩 높다.

cat_amt_z                                                    \
             count      mean       std       min       25%       50%   
is_fraud                                                               
0         489890.0 -0.007959  0.971321 -3.349764 -0.539615 -0.229063   
1           1244.0  3.134379  3.468328 -4.264316 -0.136667  3.471218   

                                  amt_z                                \
               75%        max     count      mean       std       min   
is_fraud                                                                
0         0.270950  18.900493  489890.0 -0.007444  0.977602 -0.922785   
1         5.303739  16.128909    1244.0  2.931319  3.120485 -0.817302   

                                                  
               25%      50%       75%        max  
is_fraud                                          
0        -0.361132 -0.17814  0.093860  63.421230  
1        -0.188017  2.09175  5.380819  16.176233

## 3) 결제 시간 Feature 생성
사기 거래는 특정 시간대에 집중되는 경향. 고객의 평소 거래 시간대와 비교하여 이상 패턴을 탐지

In [285]:
# trans_date_trans_time 컬럼에서 시간 값을 추출하여 hour 컬럼에 저장
cc_df['trans_date_trans_time']=pd.to_datetime(cc_df['trans_date_trans_time'])

In [286]:
cc_df['trans_date_trans_time']

0        2019-01-01 00:00:44
1        2019-01-01 00:12:34
2        2019-01-01 00:17:16
3        2019-01-01 00:20:15
4        2019-01-01 00:23:41
                 ...        
491129   2020-12-31 23:56:48
491130   2020-12-31 23:56:57
491131   2020-12-31 23:59:09
491132   2020-12-31 23:59:15
491133   2020-12-31 23:59:34
Name: trans_date_trans_time, Length: 491134, dtype: datetime64[ns]

In [287]:
cc_df['hour']=cc_df['trans_date_trans_time'].dt.hour

In [288]:
# 결제시간을 morning, afternoon, night, evening으로 분류
def hour_func(x):
  if (x >= 6) & (x<12):
    return 'moring'
  elif (x >= 12) & (x < 18):
    return 'afternoon'
  elif (x >= 18) & (x < 23):
    return 'night'
  else:
    return 'evening'

cc_df['hour_cat']=cc_df['hour'].apply(hour_func)

In [289]:
cc_df.head()

trans_date_trans_time               cc_num     category     amt gender  \
0   2019-01-01 00:00:44         630423337322  grocery_pos  107.23      F   
1   2019-01-01 00:12:34  4956828990005111019  grocery_pos   44.71      M   
2   2019-01-01 00:17:16      180048185037117  grocery_net   46.28      F   
3   2019-01-01 00:20:15      374930071163758  grocery_pos   64.09      M   
4   2019-01-01 00:23:41     2712209726293386     misc_pos   25.58      F   

       lat      long  city_pop         dob  merch_lat  merch_long  is_fraud  \
0  48.8878 -118.2105       149  1978-06-21  49.159047 -118.186462         0   
1  40.6747  -74.2239    124967  1980-12-21  40.079588  -74.848087         0   
2  40.6152  -74.4150     71485  1974-07-19  40.021888  -74.228188         0   
3  42.2203  -83.3583     31515  1971-11-05  42.360426  -83.552316         0   
4  30.4066  -91.1468    378909  1977-02-22  29.737426  -90.853194         0   

      amt_z  cat_amt_z  hour hour_cat  
0  0.321302   0.317631     0  evening  
1 -0.114637  -0.652060     0  evening  
2 -0.361802  -0.723328     0  evening  
3 -0.001304   0.211780     0  evening  
4 -0.393828  -0.575698     0  evening

In [290]:
time_dist = cc_df['hour_cat'].value_counts()

In [291]:
time_dist

hour_cat
afternoon    176801
night        146697
evening       98662
moring        68974
Name: count, dtype: int64

In [292]:
# 고객별 전체 거래 수
# 각 카드번호(cc_num)별로 그루핑, 각 카드번호 그룹 내에서 amt 컬럼의 값 개수(즉, 거래 횟수)를 셈
all_cnt = cc_df.groupby('cc_num')['amt'].count().reset_index()

In [293]:
all_cnt.columns = ['cc_num', 'total_cnt']

In [294]:
all_cnt

cc_num  total_cnt
0           503874407318       3655
1           567868110212       3644
2           571365235126       4374
3           581686439828       3653
4           630423337322       4362
..                   ...        ...
119  4792627764422477317       4357
120  4797297220948468262       3653
121  4861310130652566408       3643
122  4906628655840914250       3655
123  4956828990005111019       3657

[124 rows x 2 columns]

In [295]:
# 고객별 시간대별 거래 수
hour_cnt = cc_df.groupby(['cc_num', 'hour_cat'])['amt'].count().reset_index()

In [296]:
hour_cnt

cc_num   hour_cat   amt
0           503874407318  afternoon  1280
1           503874407318    evening   737
2           503874407318     moring   558
3           503874407318      night  1080
4           567868110212  afternoon  1228
..                   ...        ...   ...
491  4906628655840914250      night  1073
492  4956828990005111019  afternoon  1277
493  4956828990005111019    evening   784
494  4956828990005111019     moring   558
495  4956828990005111019      night  1038

[496 rows x 3 columns]

In [297]:
hour_cnt.columns = ['cc_num', 'hour_cat', 'hour_cnt']

In [298]:
 hour_cnt = hour_cnt.merge(all_cnt, on='cc_num', how='left')

In [299]:
hour_cnt

cc_num   hour_cat  hour_cnt  total_cnt
0           503874407318  afternoon      1280       3655
1           503874407318    evening       737       3655
2           503874407318     moring       558       3655
3           503874407318      night      1080       3655
4           567868110212  afternoon      1228       3644
..                   ...        ...       ...        ...
491  4906628655840914250      night      1073       3655
492  4956828990005111019  afternoon      1277       3657
493  4956828990005111019    evening       784       3657
494  4956828990005111019     moring       558       3657
495  4956828990005111019      night      1038       3657

[496 rows x 4 columns]

In [300]:
hour_cnt['hour_perc'] = hour_cnt['hour_cnt'] / hour_cnt['total_cnt']

In [301]:
hour_cnt.head(10)
# 고객 1000번이 평소 오전(0.429)에 주로 거래하는데, 갑자기 새벽(evening)에 거래한다면 이상 신호로 판단

cc_num   hour_cat  hour_cnt  total_cnt  hour_perc
0  503874407318  afternoon      1280       3655   0.350205
1  503874407318    evening       737       3655   0.201642
2  503874407318     moring       558       3655   0.152668
3  503874407318      night      1080       3655   0.295486
4  567868110212  afternoon      1228       3644   0.336992
5  567868110212    evening       820       3644   0.225027
6  567868110212     moring       529       3644   0.145170
7  567868110212      night      1067       3644   0.292810
8  571365235126  afternoon      1523       4374   0.348194
9  571365235126    evening       943       4374   0.215592

In [302]:
hour_cnt.to_pickle('./hour_cnt.pkl')

In [303]:
cc_df = cc_df.merge(hour_cnt,on=['cc_num','hour_cat'],how='left')

In [304]:
cc_df.head(10)

trans_date_trans_time               cc_num       category     amt gender  \
0   2019-01-01 00:00:44         630423337322    grocery_pos  107.23      F   
1   2019-01-01 00:12:34  4956828990005111019    grocery_pos   44.71      M   
2   2019-01-01 00:17:16      180048185037117    grocery_net   46.28      F   
3   2019-01-01 00:20:15      374930071163758    grocery_pos   64.09      M   
4   2019-01-01 00:23:41     2712209726293386       misc_pos   25.58      F   
5   2019-01-01 00:26:22     3598215285024754    grocery_pos  207.36      F   
6   2019-01-01 00:27:45      213141712584544       misc_pos  181.35      F   
7   2019-01-01 00:30:49         676372984911  gas_transport   53.60      F   
8   2019-01-01 00:31:51     4642255475285942  gas_transport  102.13      F   
9   2019-01-01 00:34:10     3553629419254918    grocery_pos   83.07      F   

       lat      long  city_pop         dob  merch_lat  merch_long  is_fraud  \
0  48.8878 -118.2105       149  1978-06-21  49.159047 -118.186462         0   
1  40.6747  -74.2239    124967  1980-12-21  40.079588  -74.848087         0   
2  40.6152  -74.4150     71485  1974-07-19  40.021888  -74.228188         0   
3  42.2203  -83.3583     31515  1971-11-05  42.360426  -83.552316         0   
4  30.4066  -91.1468    378909  1977-02-22  29.737426  -90.853194         0   
5  40.6729  -73.5365     34496  1970-10-21  40.202538  -73.859212         0   
6  32.1530  -90.1217     19685  1984-12-24  31.939812  -90.726209         0   
7  39.7562  -98.4462       242  1987-07-18  40.332519  -99.062962         0   
8  38.9999 -109.6150        46  1987-04-23  39.807313 -109.348294         0   
9  48.3400 -122.3456        85  1984-09-01  48.682111 -122.719904         0   

      amt_z  cat_amt_z  hour hour_cat  hour_cnt  total_cnt  hour_perc  
0  0.321302   0.317631     0  evening       857       4362   0.196470  
1 -0.114637  -0.652060     0  evening       784       3657   0.214383  
2 -0.361802  -0.723328     0  evening       952       4382   0.217252  
3 -0.001304   0.211780     0  evening       779       3647   0.213600  
4 -0.393828  -0.575698     0  evening       887       4372   0.202882  
5  1.012426   0.316455     0  evening       744       3650   0.203836  
6  0.766191   1.193414     0  evening       901       4371   0.206131  
7 -0.035593  -0.387377     0  evening       770       3649   0.211017  
8  0.296075   2.797777     0  evening       900       4386   0.205198  
9  0.293265  -0.800114     0  evening       869       4381   0.198357

In [305]:
cc_df.drop(['trans_date_trans_time', 'hour', 'hour_cat'], axis =1 , inplace = True)

##4) 거리 feature 분석
고객의 평소 활동 반경을 벗어난 거래는 사기 가능성이 높습니다. 지리적 위치 정보를 활용하여 이상 거래를 탐지

In [306]:
!pip install geopy

In [307]:
from geopy.distance import distance

In [308]:
# 위도와 경도 값으로 두 지점 사이의 거리를 구하기
# 고객의 위치와 상점의 위치 사이의 거리를 계산하여 distance 컬럼에 저장
cc_df['distance'] = cc_df.apply(lambda x: distance((x['lat'], x['long']), (x['merch_lat'], x['merch_long'])).km, axis = 1)

In [309]:
# 거리 Z-Score

dist_info = cc_df.groupby('cc_num')['distance'].agg(['mean', 'std']).reset_index()

In [310]:
dist_info.to_pickle('./dist_info.pkl')

In [311]:
cc_df=cc_df.merge(dist_info, on='cc_num', how='left')

In [312]:
cc_df.shape

(491134, 19)

In [313]:
cc_df['dist_z']= (cc_df['distance'] - cc_df['mean']) / cc_df['std']

In [314]:
fraud_dist = cc_df.groupby('is_fraud')['dist_z'].describe()

In [315]:
fraud_dist
# 사기 거래의 평균 거리 Z-Score가 0..014로, 평소보다 훨씬 먼 곳에서 거래가 발생

count      mean       std       min       25%       50%  \
is_fraud                                                               
0         489890.0 -0.000037  0.999918 -2.729564 -0.712473  0.078605   
1           1244.0  0.014573  0.982953 -2.585458 -0.703523  0.093968   

               75%       max  
is_fraud                      
0         0.764105  2.323064  
1         0.775749  2.133614

In [316]:
## 5) 나이 Feature 생성
cc_df['dob']=pd.to_datetime(cc_df['dob']).dt.year

In [317]:
cc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 491134 entries, 0 to 491133
Data columns (total 20 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   cc_num      491134 non-null  int64  
 1   category    491134 non-null  object 
 2   amt         491134 non-null  float64
 3   gender      491134 non-null  object 
 4   lat         491134 non-null  float64
 5   long        491134 non-null  float64
 6   city_pop    491134 non-null  int64  
 7   dob         491134 non-null  int32  
 8   merch_lat   491134 non-null  float64
 9   merch_long  491134 non-null  float64
 10  is_fraud    491134 non-null  int64  
 11  amt_z       491134 non-null  float64
 12  cat_amt_z   491134 non-null  float64
 13  hour_cnt    491134 non-null  int64  
 14  total_cnt   491134 non-null  int64  
 15  hour_perc   491134 non-null  float64
 16  distance    491134 non-null  float64
 17  mean        491134 non-null  float64
 18  std         491134 non-null  float64
 19  di

## 5) 원-핫 인코딩

In [318]:
cc_df = cc_df_encoded = pd.get_dummies(cc_df, columns=['category', 'gender'], drop_first=True)

In [319]:
cc_df

cc_num     amt      lat      long  city_pop   dob  \
0              630423337322  107.23  48.8878 -118.2105       149  1978   
1       4956828990005111019   44.71  40.6747  -74.2239    124967  1980   
2           180048185037117   46.28  40.6152  -74.4150     71485  1974   
3           374930071163758   64.09  42.2203  -83.3583     31515  1971   
4          2712209726293386   25.58  30.4066  -91.1468    378909  1977   
...                     ...     ...      ...       ...       ...   ...   
491129     6011109736646996  134.26  34.2651  -77.8670    186140  1983   
491130      213112402583773   25.49  34.0326  -82.2027      1523  1984   
491131     3556613125071656  111.84  29.0393  -95.4401     28739  1999   
491132     6011724471098086   86.88  46.1966 -118.9017      3684  1981   
491133     4170689372027579   38.13  35.6665  -97.4798    116001  1993   

        merch_lat  merch_long  is_fraud     amt_z  cat_amt_z  hour_cnt  \
0       49.159047 -118.186462         0  0.321302   0.317631       857   
1       40.079588  -74.848087         0 -0.114637  -0.652060       784   
2       40.021888  -74.228188         0 -0.361802  -0.723328       952   
3       42.360426  -83.552316         0 -0.001304   0.211780       779   
4       29.737426  -90.853194         0 -0.393828  -0.575698       887   
...           ...         ...       ...       ...        ...       ...   
491129  34.853497  -78.664158         0  0.360311   1.450980       914   
491130  35.008839  -81.475156         0 -0.173577  -0.128877       909   
491131  29.661049  -96.186633         0  0.364619   1.163822       587   
491132  46.658340 -119.715054         0 -0.015209   0.440137       796   
491133  36.210097  -97.036372         0 -0.152588  -0.380698       734   

        total_cnt  hour_perc    distance       mean        std    dist_z  \
0            4362   0.196470   30.216618  71.656621  28.090646 -1.475224   
1            3657   0.214383   84.714605  74.811123  28.675031  0.345370   
2            4382   0.217252   67.768167  75.617531  28.784606 -0.272693   
3            3647   0.213600   22.322745  74.706461  28.711493 -1.824486   
4            4372   0.202882   79.398244  79.334924  29.620117  0.002138   
...           ...        ...         ...        ...        ...       ...   
491129       4369   0.209201   98.043837  77.251486  29.270739  0.710346   
491130       4380   0.207534  127.240424  77.491795  29.899826  1.663843   
491131       3661   0.160339  100.023736  79.172599  30.051452  0.693848   
491132       3651   0.218022   80.887812  72.656420  28.552151  0.288293   
491133       3651   0.201041   72.380990  77.087902  29.521711 -0.159439   

        category_food_dining  category_gas_transport  category_grocery_net  \
0                      False                   False                 False   
1                      False                   False                 False   
2                      False                   False                  True   
3                      False                   False                 False   
4                      False                   False                 False   
...                      ...                     ...                   ...   
491129                 False                   False                 False   
491130                 False                   False                 False   
491131                 False                   False                 False   
491132                 False                   False                 False   
491133                 False                   False                 False   

        category_grocery_pos  category_health_fitness  category_home  \
0                       True                    False          False   
1                       True                    False          False   
2                      False                    False          False   
3                       True                    False          False   
4                      False     

In [320]:
cc_df.drop('cc_num',axis=1, inplace=True)

In [321]:
cc_df

amt      lat      long  city_pop   dob  merch_lat  merch_long  \
0       107.23  48.8878 -118.2105       149  1978  49.159047 -118.186462   
1        44.71  40.6747  -74.2239    124967  1980  40.079588  -74.848087   
2        46.28  40.6152  -74.4150     71485  1974  40.021888  -74.228188   
3        64.09  42.2203  -83.3583     31515  1971  42.360426  -83.552316   
4        25.58  30.4066  -91.1468    378909  1977  29.737426  -90.853194   
...        ...      ...       ...       ...   ...        ...         ...   
491129  134.26  34.2651  -77.8670    186140  1983  34.853497  -78.664158   
491130   25.49  34.0326  -82.2027      1523  1984  35.008839  -81.475156   
491131  111.84  29.0393  -95.4401     28739  1999  29.661049  -96.186633   
491132   86.88  46.1966 -118.9017      3684  1981  46.658340 -119.715054   
491133   38.13  35.6665  -97.4798    116001  1993  36.210097  -97.036372   

        is_fraud     amt_z  cat_amt_z  hour_cnt  total_cnt  hour_perc  \
0              0  0.321302   0.317631       857       4362   0.196470   
1              0 -0.114637  -0.652060       784       3657   0.214383   
2              0 -0.361802  -0.723328       952       4382   0.217252   
3              0 -0.001304   0.211780       779       3647   0.213600   
4              0 -0.393828  -0.575698       887       4372   0.202882   
...          ...       ...        ...       ...        ...        ...   
491129         0  0.360311   1.450980       914       4369   0.209201   
491130         0 -0.173577  -0.128877       909       4380   0.207534   
491131         0  0.364619   1.163822       587       3661   0.160339   
491132         0 -0.015209   0.440137       796       3651   0.218022   
491133         0 -0.152588  -0.380698       734       3651   0.201041   

          distance       mean        std    dist_z  category_food_dining  \
0        30.216618  71.656621  28.090646 -1.475224                 False   
1        84.714605  74.811123  28.675031  0.345370                 False   
2        67.768167  75.617531  28.784606 -0.272693                 False   
3        22.322745  74.706461  28.711493 -1.824486                 False   
4        79.398244  79.334924  29.620117  0.002138                 False   
...            ...        ...        ...       ...                   ...   
491129   98.043837  77.251486  29.270739  0.710346                 False   
491130  127.240424  77.491795  29.899826  1.663843                 False   
491131  100.023736  79.172599  30.051452  0.693848                 False   
491132   80.887812  72.656420  28.552151  0.288293                 False   
491133   72.380990  77.087902  29.521711 -0.159439                 False   

        category_gas_transport  category_grocery_net  category_grocery_pos  \
0                        False                 False                  True   
1                        False                 False                  True   
2                        False                  True                 False   
3                        False                 False                  True   
4                        False                 False                 False   
...                        ...                   ...                   ...   
491129                   False                 False                 False   
491130                   False                 False                 False   
491131                   False                 False                 False   
491132                   False                 False                 False   
491133                   False                 False                 False   

        category_health_fitness  category_home  category_kids_pets  \
0                         False          False               False   
1                         False          False               False   
2                         False          False               False   
3                         False          False               False   
4                    